Added support for sha256 checksums

woodb opened the pull request:
This PR adds support for sha256 checksums for io.js and node.js.

ljharb commented:
None of these sha-256 utilities are part of POSIX, nor natively available on a Mac, nor available on travis-ci. Merging this PR as-is would break 100% of the supported use cases of `nvm`.

If you could rework this to provide optional checking when a utility was available, printing to stderr when one was not, but otherwise continuing, I would be very likely to include it.

See #664 for a previous attempt.

woodb commented:
> None of these sha-256 utilities are part of POSIX

Many Linux distributions have a `sha256sum` available to them, such as [Ubuntu in coreutils](https://help.ubuntu.com/community/HowToSHA256SUM). 

> None of these sha-256 utilities are part of [...] nor natively available on a Mac. 

I did run these tests on MacOS 10.11.2 without failure to checksum (with GNU `sha256sum` installed as `gsha256sum`). I was glad to find that the native `shasum` supported this. I have added support for that to this PR and noted credit to @DomT4 as best I could.

> None of these sha-256 utilities are part of [...] travis-ci

The original commit (https://github.com/woodb/nvm/commit/560c8c9500cb2d920527d6889e37f4d3fd1fb361) [did pass the test suite on TravisCI](https://travis-ci.org/creationix/nvm/builds/104383980)

> rework this to provide optional checking when a utility was available, printing to stderr when one was not, but otherwise continuing

I have added this, also (with stderr output).

woodb commented:
:warning: I'm working through the issues produceable only on the TravisCI environment currently and **will be force pushing** to this branch.

ljharb commented:
Great, thanks - force pushing to a branch is always the right way to go :-)

DomT4 commented:
Hey, Thanks for picking this back up @woodb! No worries on any credit or such. It'll be great to see this done.

> I was glad to find that the native shasum supported this.

The system OpenSSL, as old and decrepit as it is these days, is also capable of processing SHA1 and SHA256 checksums, so that's another potential fallback. I can't immediately think of a system off the top of my head that wouldn't answer to either `sha256sum`, `shasum` or `openssl`, though I'm sure one exists.

woodb commented:
Hey @DomT4, thanks for the note! I think I've got a working solution pushed up now, please feel free to review.

ljharb commented on nvm.sh:
```diff
@@ -759,15 +759,37 @@ nvm_ls_remote_index_tab() {
+      NVM_CHECKSUM="$(shasum -a 256 "$1" | awk '{print $1}')"
+    elif nvm_has "sha256" && ! nvm_is_alias "sha256"; then
+      NVM_CHECKSUM="$(sha256 -q "$1" | awk '{print $1}')"
+    elif nvm_has "gsha256sum" && ! nvm_is_alias "gsha256sum"; then
```
is this a common alias for this program, or is this just the one you happen to use?

woodb replied:
Homebrew on OS X commonly [prefixes GNU utilities with `g`](https://github.com/Homebrew/homebrew/blob/bccdd2869233109e61d02f8a0f5081e1dd6aba85/Library/Formula/coreutils.rb#L67), including `sha256sum`.

ljharb replied:
gotcha, thanks

DomT4 replied:
MacPorts does the same thing as well AFAIK. Not sure about Fink.

ljharb commented on nvm.sh:
```diff
@@ -759,15 +759,37 @@ nvm_ls_remote_index_tab() {
+    elif nvm_has "bssl" && ! nvm_is_alias "bssl"; then
+      NVM_CHECKSUM="$(bssl sha256sum "$1" | awk '{print $1}')"
+    else
+      echo "Unaliased sha256sum, sha256, or gsha256sum not found." >&2
```
please update this list to include all utilities checked for

woodb replied:
Thanks, I've updated this list in the most recent commit (https://github.com/creationix/nvm/commit/f73bfb65780709b181cc4f53bff554f67f26a2b3).

ljharb commented:
Tested this locally and this seems to work great. There's some weirdness around the "from source" fallback when i force the sums to fail but that's definitely a previous issue.

Just a few comments and I think this is good to go!

woodb commented:
Alrighty, I think we should be good to go with https://github.com/woodb/nvm/commit/f73bfb65780709b181cc4f53bff554f67f26a2b3 :+1:

DomT4 commented:
This is great. Thank you for getting this done! :bow:
